presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -258,35 +265,48 @@ private void processNewSources()
     {
         checkLockNotHeld("Can not process while holding the driver lock");
 
+        // if the driver is blocked we don't need to continue
+        if (!previouslyReturnedBlockedFuture.isDone()) {
+            return previouslyReturnedBlockedFuture;
+        }
+
         Optional<ListenableFuture<?>> result = tryWithLock(100, TimeUnit.MILLISECONDS, () -> {
             ListenableFuture<?> driverBlockedFuture = processInternal();
-            return wakeUpOnRevokeRequest(driverBlockedFuture);
+            previouslyReturnedBlockedFuture = wakeUpOnRevokeRequest(driverBlockedFuture);
+            return previouslyReturnedBlockedFuture;
         });
         return result.orElse(NOT_BLOCKED);
     }
 
     private ListenableFuture<?> wakeUpOnRevokeRequest(ListenableFuture<?> driverBlockedFuture)
     {
+        verify(previouslyReturnedBlockedFuture.isDone(), "previouslyReturnedBlockedFuture must be done");
+
         if (driverBlockedFuture.isDone()) {
             // driver is not blocked; just return completed future
             return driverBlockedFuture;
         }
 
-        ImmutableList<SettableFuture<?>> revokingRequestedFutures = operators.stream()
+        // check to see whether revoking is requested for any of the operators
+        boolean isRevokingRequested = operators.stream()
                 .map(Operator::getOperatorContext)
-                .map(OperatorContext::getMemoryRevokingRequestedFuture)
-                .collect(toImmutableList());
-        Optional<SettableFuture<?>> doneRevokingRequestedFuture = revokingRequestedFutures.stream().filter(Future::isDone).findAny();
-        if (doneRevokingRequestedFuture.isPresent()) {
-            // revoking already requested for one of operators; return completed future
-            return doneRevokingRequestedFuture.get();
-        }
-
-        // unblock as soon as driver is unblocked or we get revoking request for any of operators
-        ImmutableList.Builder<ListenableFuture<?>> futures = ImmutableList.builder();
-        futures.add(driverBlockedFuture);
-        futures.addAll(revokingRequestedFutures);
-        return firstFinishedFuture(futures.build());
+                .filter(OperatorContext::isMemoryRevokingRequested)
```

Comment:
we need to also check if we are not doing actual revoking. If operator is in `revokingOperators` map it should not unblock driver until revoking completes.

Also - do we need this section at all? Seems like `setMemoryRevokingRequestedListener` below should do the job and call listener immediately if  revoking was already requested for `OperatorContext`. The filtering using `revokingOperators` is still needed though.